### ISSUE TITLE
NUMERIC vectors

### ISSUE THREAD

**Issue by bgurmendi** [pjungwir/floatvec#1]
will work this extension for numeric type arrays?

**Comment by pjungwir** [comment]
No, sorry! It could be extended without too much effort I suppose. Let me know if you're still interested in using something like that.

**Comment by msqr** [comment]
This is a great extension, and I, too am looking for `NUMERIC` support. I started working on this in [SolarNetwork/floatvec](https://github.com/SolarNetwork/floatvec/tree/feature/numeric). I had a couple of questions I hoped @pjungwir you could help me with?

1. I wanted a ["zero" Numeric cached global ](https://github.com/SolarNetwork/floatvec/commit/a8608969f363a4de46ca96530a7a5ae4b00d3e3c#diff-3f0e35fe7bd32fcae40836d2e831629f7a0782c5a9cdac388831f5e7869def89R23-R43)to help with performance of [testing for zero before dividing](https://github.com/SolarNetwork/floatvec/commit/45be9b1cac86452a7798f62e13cc33dc90dce1db#diff-dc4b0bc3b496d006928f7ce0fbe8a1f3c4d8008ea072aad894a5bed0aa11e555R93). As I am a newbie to Postgres, do you think what I've done is a suitable way to achieve this? I see that Postgres has [a private way of doing a similar thing](https://github.com/postgres/postgres/blob/e54a758d24dab056bb7f50d26c57a3c8761cc44a/src/backend/utils/adt/numeric.c#L416-L418).
2. Postgres 14 added support for `Infinity` values with the `NUMERIC` type. I implemented support for that in `vec_div`, returning `NaN` for Postgres < 14. I couldn't find a way to write test cases that were Postgres version-specific, however. [These tests](https://github.com/SolarNetwork/floatvec/commit/45be9b1cac86452a7798f62e13cc33dc90dce1db#diff-8bd71f53a752d13e9803ff92c7eb2a4d4aa598748032bf1a9493e49c9ebbea82R234-R247) are written for Postgres < 14, and thus fail when running in 14.

I plan to finish implementing `NUMERIC` support for all these functions. If you like the look of the work, I can submit a PR. Thanks for your time!

**Comment by pjungwir** [comment]
> 1. I wanted a "zero" Numeric cached global to help with performance of testing for zero before dividing. As I am a newbie to Postgres, do you think what I've done is a suitable way to achieve this?

I looked at your code and it seems fine to me. But you have a `FIXME` comment, so it's not working? I can help you debug it if you like. I agree having a constant zero seems like a good idea.

> 2. I couldn't find a way to write test cases that were Postgres version-specific

I'd like to switch the tests to `pg_regress` and the PGXN testing framework, which is more standard. It's used for Postgres itself and its own `contrib` extensions. I've got a local commit for that already for [aggs_for_vecs](https://github.com/pjungwir/aggs_for_vecs). We could do it here too (or I'm more inclined to just pull everything into `aggs_for_vecs`). Once I do that, we can write two versions of the expected output for a given test file, and Postgres will pass the test if *either* of them match. For example you can see this in Postgres's own [xml test](https://github.com/postgres/postgres/blob/master/src/test/regress/sql/xml.sql) with results [with](https://github.com/postgres/postgres/blob/master/src/test/regress/expected/xml_2.out) and [without](https://github.com/postgres/postgres/blob/master/src/test/regress/expected/xml_1.out) libxml.

**Comment by msqr** [comment]
Ah great; the FIXME was a reminder to validate the approach, it is actually working for me and the tests all pass so I'll remove that comment. Per our [discussion on aggs_for_vecs](https://github.com/pjungwir/aggs_for_vecs/issues/5#issuecomment-983223689) I will copy this work over to that project, having everything there will make it easier for me as well!
